developer/frameworks/articles/user-associated-data/03-gdpr-deletion-features.markdown

```diff
@@ -0,0 +1,216 @@
+which can contain both files and other folders. The non-container entities can
+only be children in the hierarchy. An example is a file in a file
+system.
+
+The `UADHierarchyDeclaration` provides some other methods for display purposes.
+
+- A label for the hierarchy is provided through the `getEntitiesTypeLabel`
+    method.
+
+- If any additional information is to be displayed in the table for any of the
+    entity types in addition to the count, those column names should be returned
+    by `getExtraColumnNames`. This is optional.
+
+See [`DLUADHierarchyDeclaration`](https://github.com/liferay/liferay-portal/blob/master/modules/apps/document-library/document-library-uad/src/main/java/com/liferay/document/library/uad/display/DLUADHierarchyDeclaration.java) for an example.
+
+### Add methods to UADDisplay
+
+Each type involved in the hierarchy should implement some additional methods in `UADDisplay`.
+
+These methods must be implemented by containers and non-containers alike, in the
+`*UADDisplay` class:
+
+`getName` returns a display name for the given entity.
+
+`getParentContainerClass` returns the class of the type that contains this type.
+It can return itself (for example, a folder can contain a folder).
+
+`getParentContainerId` returns the primary key of the container that contains
+the entity passed to this method.
+
+`isUserOwned` returns whether or not the given entity is owned by the user
+
+Additionally, implement this method in the `*UADDisplay` class of all containers:
+
+`getTopLevelContainer` is used to derive the count of how many user-owned
+entities are contained inside a given container's tree.
+
+It answers the question "which type `T` ancestor of `childObject` is an
+immediate child of the container identified by `parentContainerClass` and
+`parentContainerId`?" The method may return null, if `childObject` is not a
+child of the parent container at all. It's method is the most complicated to
+implement and requires some consideration for each case. Refer to the test case
+for examples of the requirements used for `DLFolder`:
+[DLFolderUADDisplayTest#testGetTopLevelContainer](https://github.com/liferay/liferay-portal/blob/c8f78609353d6a83a0b755b0bbf93764959821ee/modules/apps/document-library/document-library-uad-test/src/testIntegration/java/com/liferay/document/library/uad/display/test/DLFolderUADDisplayTest.java#L67)
+
+See the actual implementation for `DLFolder` in [DLFolderUADDisplay#getTopLevelContainer](https://github.com/liferay/liferay-portal/blob/c8f78609353d6a83a0b755b0bbf93764959821ee/modules/apps/document-library/document-library-uad/src/main/java/com/liferay/document/library/uad/display/DLFolderUADDisplay.java#L105).
+
+The method returns either `null` or the `childObject` (could be any `Object`)
+representing the top level container that holds entities for the given User in
+this particular UAD task, so that the User's entities inside the container can
+be counted.
```

Comment:
I don't think this paragraph is accurate. May have been my mistake.

```markdown
The method returns either `null` or the container object of type T that is the top level container of the `childObject` (which could be any type of object that is a part of the hierarchy). This container does not necessarily have to be owned by the user, but is understood to contain data related to the user. This information is used to count how much user data is contained inside the container designated by `parentContainerClass` and `parentContainerId`.
```

Let me know if that seems ok to you, @rbohl

Reply:
Nah, I think I tried to read the code myself here, so it's my mistake. :grin: